src/common/admin_socket.cc

```diff
@@ -693,6 +695,281 @@ class GetdescsHook : public AdminSocketHook {
+  ADD_SIGNAL(SIGFPE),
+  ADD_SIGNAL(SIGILL),
+  ADD_SIGNAL(SIGINT),
+  ADD_SIGNAL(SIGSEGV),
+  ADD_SIGNAL(SIGTERM),
+#ifndef WIN32
+  ADD_SIGNAL(SIGTRAP),
+  ADD_SIGNAL(SIGHUP),
+  ADD_SIGNAL(SIGBUS),
+  ADD_SIGNAL(SIGQUIT),
+  ADD_SIGNAL(SIGKILL),
+  ADD_SIGNAL(SIGUSR1),
+  ADD_SIGNAL(SIGUSR2),
+  ADD_SIGNAL(SIGPIPE),
+  ADD_SIGNAL(SIGALRM),
+  ADD_SIGNAL(SIGCHLD),
+  ADD_SIGNAL(SIGCONT),
+  ADD_SIGNAL(SIGSTOP),
+  ADD_SIGNAL(SIGTSTP),
+  ADD_SIGNAL(SIGTTIN),
+  ADD_SIGNAL(SIGTTOU),
+#endif
+  // Add more signals as needed...
+};
+
+#undef ADD_SIGNAL
+
+class RaiseHook: public AdminSocketHook {
+  struct Killer {
+    CephContext* m_cct;
+    pid_t pid;
+    int signal;
+    ceph::coarse_mono_clock::time_point due;
+
+    std::string describe()
+    {
+      using std::chrono::duration_cast;
+      using std::chrono::seconds;
+      auto remaining = (due - coarse_mono_clock::now());
+      return fmt::format(
+        "pending signal ({}) due in {}", 
+        strsignal(signal), 
+        duration_cast<seconds>(remaining));
+    }
+
+    bool cancel()
+    {
+#   ifndef WIN32
+      int wstatus;
+      int status;
+      if (0 == (status = waitpid(pid, &wstatus, WNOHANG))) {
+        status = kill(pid, SIGKILL);
+        if (status) {
+          ldout(m_cct, 5) << __func__ << "couldn't kill the killer. Error: " << strerror(errno) << dendl;
+          return false;
+        }
+        while (pid == waitpid(pid, &wstatus, 0)) {
+          if (WIFEXITED(wstatus)) {
+            return false;
+          }
+          if (WIFSIGNALED(wstatus)) {
+            return true;
+          }
+        }
+      }
+      if (status < 0) {
+        ldout(m_cct, 5) << __func__ << "waitpid(killer, NOHANG) returned " << status << "; " << strerror(errno) << dendl;
+      } else {
+        ldout(m_cct, 20) << __func__ << "killer process " << pid << "\"" << describe() << "\" reaped. "
+                         << "WIFEXITED: " << WIFEXITED(wstatus)
+                         << "WIFSIGNALED: " << WIFSIGNALED(wstatus)
+                         << dendl;
+      }
+#   endif
+      return false;
+    }
+
+    static std::optional<Killer> fork(CephContext *m_cct, int signal_to_send, double delay) {
+#   ifndef WIN32
+      pid_t victim = getpid();
+      auto until = ceph::coarse_mono_clock::now() + ceph::make_timespan(delay);
+
+      int fresult = ::fork();
+      if (fresult < 0) {
+        ldout(m_cct, 5) << __func__ << "couldn't fork the killer. Error: " << strerror(errno) << dendl;
+        return std::nullopt;
+      }
+
+      if (fresult) {
+        // this is parent
+        return {{m_cct, fresult, signal_to_send, until}};
+      }
+
+      const auto poll_interval = ceph::make_timespan(0.1);
+      auto remaining = (until - ceph::coarse_mono_clock::now());
+      do {
+        using std::chrono::duration_cast;
+        using std::chrono::nanoseconds;
+        std::this_thread::sleep_for(duration_cast<nanoseconds>(std::min(remaining, poll_interval)));
```

Comment:
We use this polling due to the pid check below? Would it not suffice to sleep for the entire duration and then check the pids post the sleep? (pid reuse issue?)

Reply:
I wanted to avoid having the forked child pin all that memory beyond the parent's lifetime to only find out about it after the timeout, so this loop was the simplest mitigation I could think of.
The timeout could be arbitrarily large, so some kind of mitigation is required.

Reply:
Have you considered using prctl w/ PR_SET_PDEATHSIG?

Reply:
No, I didn't know about this one, thanks!
I mean, I suspected there was a more elegant way to handle this case, but again, decided to go with a simpler solution thinking that it would be easier to manage.

For example, this solution you suggested which I liked, is not available on MacOS :/.

Reply:
We could use the prctl for Linux and fallback to polling for other systems.

Reply:
Which mean #ifdef conditionals, so, I'm probably fine with using a single solution that works across 🤷